src/map.c

```diff
@@ -37,77 +42,101 @@ SEXP call_loop(SEXP env, SEXP call, int n, SEXPTYPE type, int force_args,
     UNPROTECT(1);
   }
 
-  INTEGER(i_val)[0] = 0;
+  *p_i = 0;
   cli_progress_done(bar);
 
   UNPROTECT(2);
   return out;
 }
 
-SEXP map_impl(SEXP env, SEXP type_, SEXP progress, SEXP error_call) {
-  static SEXP f_call = NULL;
-  if (f_call == NULL) {
-    SEXP x = Rf_install(".x");
-    SEXP f = Rf_install(".f");
-    SEXP i = Rf_install("i");
+SEXP map_impl(SEXP env,
+              SEXP ffi_type,
+              SEXP progress,
+              SEXP ffi_n,
```

Comment:
We use this `ffi_` prefix in vctrs/rlang anytime we have an argument that comes from R that we need to massage into the right form for use in C. We prefer to give the C side the "better" name by prefixing the R side object with `ffi_`

Reply:
I just read encountered this `ffi_` stuff in vctrs. Does it stand for something?

Reply:
Foreign function interface! https://en.wikipedia.org/wiki/Foreign_function_interface

In vctrs and rlang we go a step further by using `ffi_` in the bridge function's name too (i.e. then function itself on the C side would be `ffi_map()`). This works very well if you have a function that you'd like to give the same name on the R and C side, but you have to deal with some fiddly bits when going from the R function to the C function. For instance, in vctrs we have:
- R level `vec_as_names()`
- C level `ffi_vec_as_names(r_obj* names, r_obj* repair, r_obj* ffi_quiet, r_obj* frame)`. Deals with turning R level objects into C friendly `opts` struct
- C level `vec_as_names(r_obj* names, const struct name_repair_opts* opts)` which we call elsewhere in our C code